src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/util.cpp

```diff
@@ -0,0 +1,90 @@
+// Copyright (C) 2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "util.hpp"
+
+#include "../../logging.hpp"
+#include "openvino/op/ops.hpp"
+#include "openvino/pass/pattern/op/label.hpp"  // any_input
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "openvino/util/common_util.hpp"
+
+namespace ov {
+namespace npuw {
+namespace patterns {
+namespace util {
+
+namespace opp = ov::pass::pattern;
+
+//
+// FROM:
+//     Param [1x4x1023x64] -> ShapeOf -> Gather -> ...
+//
+// TO:
+//     Const [4] -> Gather -> ...
+//
+ShapeOfToConst::ShapeOfToConst(const std::shared_ptr<ov::Model>& model) {
+    auto key_val = opp::wrap_type<ov::op::v0::Parameter>();
+    auto shapeof = opp::wrap_type<ov::op::v3::ShapeOf>({key_val});
```

Comment:
it can be `any_input`?

pattern may end on `shapeof`?

Reply:
I don't think so. I suggest for this case we match `Param->ShapeOf` exactly

Reply:
Removed Gather, now pattern ends on ShapeOf